mindone/models/dit.py

```diff
@@ -0,0 +1,709 @@
+            self.patches_resolution: Optional[Tuple] = tuple([s // p for s, p in zip(self.image_size, self.patch_size)])
+            self.num_patches: Optional[int] = self.patches_resolution[0] * self.patches_resolution[1]
+        else:
+            self.image_size: Optional[Tuple] = None
+            self.patches_resolution: Optional[Tuple] = None
+            self.num_patches: Optional[int] = None
+        self.embed_dim = embed_dim
+        self.proj = nn.Conv2d(
+            in_chans, embed_dim, kernel_size=patch_size, stride=patch_size, pad_mode="pad", has_bias=bias
+        )
+
+    def construct(self, x: Tensor) -> Tensor:
+        b, c, h, w = x.shape
+        if self.image_size is not None:
+            assert (h, w) == (
+                self.image_size[0],
+                self.image_size[1],
+            ), f"Input height and width ({h},{w}) doesn't match model ({self.image_size[0]},{self.image_size[1]})."
+        x = self.proj(x)
+        x = ops.reshape(x, (b, self.embed_dim, -1))  # B Ph*Pw C
+        x = ops.transpose(x, (0, 2, 1))
+        return x
+
+
+class Mlp(nn.Cell):
+    def __init__(
+        self,
+        in_features: int,
+        hidden_features: Optional[int] = None,
+        out_features: Optional[int] = None,
+        act_layer: Optional[nn.Cell] = nn.GELU,
+        drop: float = 0.0,
+    ) -> None:
+        super().__init__()
+        out_features = out_features or in_features
+        hidden_features = hidden_features or in_features
+        self.fc1 = nn.Dense(in_channels=in_features, out_channels=hidden_features, has_bias=True)
+        self.act = act_layer()
+        self.fc2 = nn.Dense(in_channels=hidden_features, out_channels=out_features, has_bias=True)
+        self.drop = nn.Dropout(p=drop)
+
+    def construct(self, x: Tensor) -> Tensor:
+        x = self.fc1(x)
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+def exists(val):
+    return val is not None
+
+
+class Attention(nn.Cell):
+    def __init__(self, dim_head, attn_drop=0.0):
+        super().__init__()
+        self.softmax = ops.Softmax(axis=-1)
+        self.transpose = ops.Transpose()
+        self.scale = dim_head**-0.5
+        self.attn_drop = nn.Dropout(p=attn_drop)
+
+    def construct(self, q, k, v, mask):
+        sim = ops.matmul(q, self.transpose(k, (0, 2, 1))) * self.scale
+
+        if exists(mask):
+            mask = self.reshape(mask, (mask.shape[0], -1))
+            if sim.dtype == ms.float16:
+                finfo_type = np.float16
+            else:
+                finfo_type = np.float32
+            max_neg_value = -np.finfo(finfo_type).max
+            mask = mask.repeat(self.heads, axis=0)
+            mask = ops.expand_dims(mask, axis=1)
+            sim.masked_fill(mask, max_neg_value)
+
+        # use fp32 for exponential inside
+        attn = self.softmax(sim.astype(ms.float32)).astype(v.dtype)
```

Comment:
is softmax also computed in fp32 in original impl?

Reply:
In [DiT torch implementation](https://github.com/facebookresearch/DiT/blob/main/models.py), they used `timm.vision_transformer.Attention`. The code implies they are not forcing fp32 for softmax computation: https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/vision_transformer.py#L57

```python
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, self.head_dim).permute(2, 0, 3, 1, 4)
        q, k, v = qkv.unbind(0)
        q, k = self.q_norm(q), self.k_norm(k)

        if self.fused_attn:
            x = F.scaled_dot_product_attention(
                q, k, v,
                dropout_p=self.attn_drop.p if self.training else 0.,
            )
        else:
            q = q * self.scale
            attn = q @ k.transpose(-2, -1)
            attn = attn.softmax(dim=-1)
            attn = self.attn_drop(attn)
            x = attn @ v

        x = x.transpose(1, 2).reshape(B, N, C)
        x = self.proj(x)
        x = self.proj_drop(x)
        return x

```